### PR TITLE
Add support for caching, and retrieving individual objects from cache

### PR DESCRIPTION
**Pull request by lumisota**

### CODE REVIEW

File: ietfdata/datatracker.py [hunk 1]
```diff
@@ -82,6 +81,16 @@ def __post_init__(self) -> None:
...
+# ---------------------------------------------------------------------------------------------------------------------------------
+# Resource type
+
+@dataclass(frozen=True)
+class Resource:
+    resource_uri : URI
+
+T = TypeVar('T', bound=Resource)
+
+
```
Comment (csperkins) [inline ietfdata/datatracker.py:93]: I like the `Resource` changes, but they seem unrelated to the cache changes. Would it make sense to split them into a separate PR?

File: ietfdata/datatracker.py [hunk 2]
```diff
@@ -745,10 +754,15 @@ class DataTracker:
     """
     A class for interacting with the IETF DataTracker.
     """
-    def __init__(self):
+    def __init__(self, cachedir: Optional[Path] = None):
```
Comment (csperkins) [inline ietfdata/datatracker.py:757]: Should this be `cache_dir`?

File: ietfdata/datatracker.py [hunk 3]
```diff
@@ -785,16 +799,48 @@ def __del__(self):
         self.session.close()
 
 
+    def _cache_filepath(self, resource_uri: URI) -> Path:
+        if self.cachedir is not None:
+            return Path(self.cachedir, resource_uri.uri[1:-1] + ".json")
+        else:
+            return Path(".")
```
Comment (csperkins) [inline ietfdata/datatracker.py:806]: This looks to be only ever called when `self.cachedir` is set. Would it be better written:
```
        assert self.cachedir is not None
        return Path(self.cachedir, resource_uri.uri[1:-1] + ".json")
```

File: ietfdata/datatracker.py (hunk 1 above)
Comment (lumisota) [reply ietfdata/datatracker.py:93]: They aren't completely unrelated; _retrieve_multi() relies on the object of type T having a resource_uri. That said, the change is probably big enough for its own PR.

File: ietfdata/datatracker.py (hunk 2 above)
Comment (lumisota) [reply ietfdata/datatracker.py]: Fixed.

File: ietfdata/datatracker.py (hunk 1 above)
Comment (csperkins) [reply ietfdata/datatracker.py:93]: Okay, makes sense. In that case, maybe leave as-is.

File: ietfdata/datatracker.py (hunk 3 above)
Comment (lumisota) [reply ietfdata/datatracker.py]: Fixed.

File: ietfdata/datatracker.py (hunk 1 above)
Comment (lumisota) [reply ietfdata/datatracker.py:93]: OK; will leave it.
